python/pyarrow/_dataset.pyx

```diff
@@ -54,12 +54,18 @@ cdef class Dataset:
         shared_ptr[CDataset] wrapped
         CDataset* dataset
 
-    def __init__(self, children, Schema schema not None):
+    cdef readonly:
+        # if a factory object has instantiated the dataset construction then
+        # hold a reference for the factory to reuse it later
+        DatasetFactory factory
```

Comment:
Can you put this in a private attribute? (or is this not visible on the python side because it's a readonly cdef attribute?)

Reply:
Yes, but I can rename it to `_factory`

Reply:
Removed the whole thing.